[API] Add dicom upload

maximemulder opened the pull request:
## Brief summary of changes

This is a port of [#6161](https://github.com/aces/Loris/pull/6161), which contains a description of the changes. Since the old PR has diverged quite a bit from the main branch, and that its author no longer works here, I found it easier to copy the changes on the current main branch and open a new PR rather than to rebase the PR. While porting, I only rewrote the old PR code whenever static analysis or practical stests showed that it was no longer in sync with the main branch (or when my IDE removed trailing spaces).

#### Testing instructions (if applicable)

- Test the four new API routes, described in section 5 of the documentation of the API v0.0.4.

### Notes for existing projects

This PR requires to run the following SQL patch `SQL/New_patches/2024-03_19-MRIUploadServerProcessRel.sql`.

MaximeMulder commented:
I must say I am not sure what I am supposed to do regarding the failed tests on files that I did not modify in the PR. Any direction @driusan (or someone else).

kongtiaowang commented:
------ --------------------------------------------------- 
  Line   php/libraries/FilesDownloadHandler.php             
 ------ --------------------------------------------------- 
  71     Parameter #1 $value of function strval expects     
         bool|float|int|resource|string|null, mixed given.  
 ------ --------------------------------------------------- 

 ------ -------------------------------------------------------------------- 
  Line   php/libraries/SiteIDGenerator.php                                   
 ------ -------------------------------------------------------------------- 
  81     Negated boolean expression is always false.                         
         💡 Because the type is coming from a PHPDoc, you can turn off this  
            check by setting treatPhpDocTypesAsCertain: false in your        
            ./test/phpstan-loris.neon.                                       
  297    Parameter #1 $value of function strval expects                      
         bool|float|int|resource|string|null, mixed given.                   
  377    Parameter #1 $value of function strval expects                      
         bool|float|int|resource|string|null, array<int,                     
         int|string>|string|null given.                                      
  387    Parameter #1 $value of function strval expects                      
         bool|float|int|resource|string|null, array<int, int|string>|string  
         given.                                                              
  400    Parameter #1 $value of function strval expects                      
         bool|float|int|resource|string|null, array<int, int|string>|string  
         given.                                                              
 ------ --------------------------------------------------------------------

MaximeMulder commented:
Hum, yeah, I already saw that.

I was just suprised to see that the errors are marked as originating from `FilesDownloadHandler.php` and `SiteIDGenerator.php`, which are not modified by this PR. I guess I must call these files from elsewhere and that the location reporting is just imprecise.

I'll tackle this tomorrow probably.

MaximeMulder commented:
I've looked at the static errors in more details, it appears to be mainly due to an upgrade of phpstan, which changes the typing of string conversions with `strval` and `(string)` from `mixed -> string` to `(bool|float|int|resource|string|null) -> string`. I still don't know why this PR triggers the error and not other ones (it did trigger on my `main` branch though), but at least the errors are clear.

However, I find the code that triggers the errors to be quite cryptic, with some lines that seem obviously wrong (at least according to typing). If someone who knows this code better than me could look into it, it would be best. In the meantime, I'm adding this to the next meeting.

MaximeMulder commented:
Ok it seems the linting errors are ~~fixed~~, silenced. @laemtl had noticed that the errors were due to a composer update on my end. The errors should still appear in any PR that updates composer and should be fixed eventually. However, this can have its own issue and be another PR.

cmadjar commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
````diff
@@ -823,90 +823,191 @@ The JSON object is of the form:
+POST /candidates/$CandID/$Visit/dicoms
+```
+
+In addition to the uploaded file, the body of the POST request should contain a
+`json` form attribute containing the following keys and values:
````
Sounds to me that there should be some text explaining that there should be one compressed file uploaded per `StudyInstanceUID` and only one file.

Example: it should be clear for the user that if a user uploads a compressed file with T1W for StudyInstanceUID 1 and then uploads a T2W for StudyInstanceUID 2, then the DICOM files for the T1W will be overwritten by the T2W.

cmadjar commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
````diff
@@ -823,90 +823,191 @@ The JSON object is of the form:
+GET /candidates/$CandID/$VisitLabel/dicoms/$Tarname/processes
+```
+
+The response contains all `mri_upload` attempts with the specified `$tarname`. And for
````
technically, there should only be one `UploadID` per `TarchiveID`. So when you overwrite, you should grep the `UploadID` associated to what is being overwritten instead of inserting a new entry in `mri_upload`. Especially since the incoming directory content will be overwritten anyway.

maximemulder replied:
I agree that there is only one MRI upload per DICOM archive, but I've seen a few cases in testing (even without this PR) where there are a few failed MRI uploads and one successful for a single file. This is how the batch uploader does it (don't know about other pipelines).

cmadjar replied:
yeah. I think in principle, there should only be one upload per tarchive but unfortunately the batch upload script is not compliant with this set up. However, for the API, I think we should make sure this is the behaviour at minimum, to avoid duplicated entries.

driusan replied:
Is this thread resolved?

cmadjar commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
````diff
@@ -823,90 +823,191 @@ The JSON object is of the form:
+containing the `MriUploadID` attribute in the request body should be sent.
+
+```
+POST /candidates/$CandID/$VisitLabel/dicoms/$Tarname/processes
````
can you trigger multiple processes on the same UploadID in the current state?

maximemulder replied:
Yes, notably if the previous uploads failed for some reason.

cmadjar commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
````diff
@@ -823,90 +823,191 @@ The JSON object is of the form:
+To obtain a specific process state, use the following:
+
+```
+GET /candidates/$CandID/$VisitLabel/dicoms/$Tarname/processes/$processid
````
in this example, would `processid` be `1` or `123`?

cmadjar commented on modules/api/php/endpoints/candidate/visit/dicoms.class.inc:
```diff
@@ -138,20 +156,167 @@ class Dicoms extends Endpoint implements \LORIS\Middleware\ETagCalculator
             $view = (new \LORIS\api\Views\Visit\Dicoms(
                 $this->_visit,
-                ...$dicomtars
+                $dicomtars
```
maybe my eyes are tricking me but why is there a special case for v0.0.3 when the default code is exactly the same as the v0.0.3?

MaximeMulder replied:
This was a remnant of the original PR, changed.

cmadjar commented on modules/api/php/endpoints/candidate/visit/dicoms.class.inc:
```diff
@@ -138,20 +156,167 @@ class Dicoms extends Endpoint implements \LORIS\Middleware\ETagCalculator
+
+        if (!isset($files['MriFile'])) {
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'The input key for the uploaded file should be MriFile.'
```
Note: most use cases we have are MRI files but technically, users could upload PET DICOM files as well, so I would tend to keep the 'MRI' string out of the docs and JSON responses and instead use the more generic term "imaging file".

MaximeMulder replied:
Hhhhm, I changed the naming of the two form-data fields to just `Json` and `File`. The documentation then goes into more details about what are the roles and structures of the expected data.

cmadjar commented on modules/api/php/endpoints/candidate/visit/dicoms.class.inc:
```diff
@@ -138,20 +156,167 @@ class Dicoms extends Endpoint implements \LORIS\Middleware\ETagCalculator
+        $file_info = new \SplFileInfo($file_path);
+        if (!$file_info->isFile()) {
+            return new \LORIS\Http\Response\JSON\InternalServerError(
+                'Upload failed (see error_log.)'
```
what `error_log` is this referring to? The apache error log? If so, most users do not have access to that file so I would probably rephrase that to contact your LORIS admin to check the error_log with a timestamp to ease parsing.

MaximeMulder replied:
I must say I am not sure which error log this is. It was part of the original PR. I removed it in favor of asking to contact the LORIS administrator. This certainly could benefit from more information but I did not find any good example to draw aspiration from in the codebase, so I stayed pretty basic.

cmadjar commented on modules/api/php/endpoints/candidate/visit/dicoms.class.inc:
```diff
@@ -138,20 +156,167 @@ class Dicoms extends Endpoint implements \LORIS\Middleware\ETagCalculator
+        } catch (\Throwable $e) {
+            error_log($e->getMessage());
+            return new \LORIS\Http\Response\JSON\InternalServerError(
+                'Database insertion failed (see error_log.)'
```
same comment regarding the error_log

MaximeMulder replied:
Same comment as the other one.

cmadjar commented on modules/api/php/endpoints/candidate/visit/dicoms.class.inc:
```diff
@@ -138,20 +156,167 @@ class Dicoms extends Endpoint implements \LORIS\Middleware\ETagCalculator
+
+        // This concatenation is here because the API part is removed from the URL by
+        // the module, which may not be a good idea.
+        $path = '/api/v0.0.4-dev/' . $request->getUri()->getPath() . '/'
```
mmm, not sure about hardcoding the API version here. @driusan @ridz1208?

In general, if users are using 0.0.3, then the endpoints would not be available and if they use 0.0.4-dev, then the endpoint would be available and the upload would work.

maximemulder replied:
This was like that in the original PR, I don't really know what I should do here TBH.

cmadjar commented on modules/imaging_browser/test/imaging_browserTest.php:
```diff
@@ -242,87 +242,86 @@ public function setUp(): void
-             'Caveat'                => 0,
-             'TarchiveSource'        => 263,
-            )
+            [
```
I don't think this file should show up in the diff of that PR. Everything seems to be changes from `array()` to `[]`. I don't remember if there is a LORIS convention on how to encode those but in any case, it should not show up in this PR.

MaximeMulder replied:
This is a remnant of the original PR, removed.

cmadjar commented on php/libraries/DicomTar.class.inc:
```diff
@@ -121,10 +121,10 @@ class DicomTar
-                    $row['repetitiontime'],
-                    $row['inversiontime'],
-                    $row['slicethickness'],
+                    \Utility::nullableStringToInt($row['echotime']),
```
ah, that answers my earlier question 🤣

MaximeMulder replied:
Whoops, that is actually a bug since those values should be floats, fixed.

cmadjar commented on src/Data/Models/DicomSeriesDTO.php:
```diff
@@ -48,21 +48,21 @@ class DicomSeriesDTO
-     * @param ?string $repetitiontime The RepetitionTime
-     * @param ?string $inversiontime  The InversionTime
-     * @param ?string $slicethickness The SliceThickness
+     * @param ?float $echotime       The EchoTime
```
shouldn't `float` be `int` in this file?

MaximeMulder replied:
Echo time is a float in our database, so I think it should be a float.

cmadjar replied:
But isn't RepetitionTime in the DB also a float for example and it is set as string in the API?

maximemulder replied:
Oh yep, but I reverted the change and made it back to a string !

The most up-to-date comment of mine is this one. It is true I did not update each point individually (probably should have, sorry !)
https://github.com/aces/Loris/pull/9154#issuecomment-2356840330

cmadjar reviewed (changes requested):
@MaximeMulder I did a first pass of review. I did not have a chance to actually test the changes as my MCIN sandbox is not letting me logging in. Have to figure out why and it could take some time so figured I would do an initial review.

MaximeMulder commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
````diff
@@ -823,90 +823,191 @@ The JSON object is of the form:
+
 ```
-GET /candidates/$CandID/$Visit/dicoms
+GET /candidates/$CandID/$VisitLabel/dicoms
````
Both are already used so this is already inconsistent unfortunately. I have a slight (not strong) preference for `$VisitLabel` so I used it here. It is also the currently more used one in the docs.

cmadjar replied:
Sounds good! Thanks!

driusan commented:
Please remove all the unrelated changes (whitespace, variable name changes, etc). If necessary you can send those in another PR. It is not feasible to review this PR with so many unrelated changed mixed together.

driusan commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
```diff
@@ -823,90 +823,191 @@ The JSON object is of the form:
-                    "SeriesNumber" : "5",
-                    "EchoTime" : "30",
-                    "RepetitionTime" : "2100",
+                    "SeriesNumber" : 5,
```
If they are not strings it's a bug and the code and should be fixed in a different PR. Numbers are not reliable in JSON since javascript only has a floating point data type to represent numbers.

MaximeMulder replied:
I think using numbers to represent floats and ints in the code is not a bug and but rather a good choice here, as we already use these representations in the database:
- For floating numbers, we go from doubles in the database to doubles in the code, so no loss of precision here.
- For integers, the values we represent are well below `Number.MAX_SAFE_INTEGER` (9007199254740991), so there should not be any problem as well.

driusan replied:
While it may be debateable whether these could have originally been safely represented as numbers, they were encoded as strings because strings faithfully represent the value given to them, while numbers can differ for various reasons related to the fact that javascript only includes floating point number (ie. values could be not representable, trailing 0s that might be significant for fixed point numbers or versions, etc) and this part of the API is already released so should not be changed.

MaximeMulder replied:
Hhhhm, this makes sense I'll change it back. Since I was not the one who originally wrote the PR I was not sure what was already part of the PR when I touched it :thinking:.

maximemulder commented:
I replied to all the comments and applied the suggested changes. 

I reverted the breaking changes to the API including floats / strings and naming. Back when I wrote them, I thought it was part of the original PR and not something that was already in LORIS.

I made the termonology more consistent by using `$VisitLabel` everywhere in the DICOM part (section 5) of the API documentation.

I also disabled my trailing spaces auto-removal and removed the associated changes.

Ready for review @cmadjar @driusan

maximemulder commented on modules/api/docs/LorisRESTAPI.md:
```diff
@@ -787,82 +787,184 @@ The JSON object is of the form:
+                {
                     "SeriesDescription" : "MPRAGE_ipat2",
-                    "SeriesNumber" : "2",
+                    "SeriesNumber" : 2,
```
Regarding this change (and others related), the current documentation is inaccurate, I can extract tis to a separate PR if wanted.

Example:
https://demo.loris.ca/api/v0.0.4-dev/candidates/587630/V1/dicoms

driusan replied:
I need to look into this to see if it's a bug in the documentation or the code. It seems weird that it's only the SeriesNumber that does it and it might be related to PHP (not LORIS) changes in the PDO over time.

It might be faster to split it out.

cmadjar reviewed (commented):
@maximemulder I re-reviewed quickly your PR. A few additional comments:

- In the description of the PR, could you add a "Notes for existing project" with the information of the SQL patches to run?
-

cmadjar commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
````diff
@@ -823,90 +823,192 @@ The JSON object is of the form:
+}
+```
+
+An optionnal header 'LORIS-Overwrite' with the value 'overwrite' can be used to
````
The `LORIS-Overwrite` feature does not work. I still get the following error if the file is already in the incoming directory:
```
<Response [409]>
{'error': 'This file has already been uploaded.'}
```

Also, wouldn't it be better to set `LORIS-Overwrite` to a boolean instead: `true` if want to overwrite existing file, `false otherwise`? And maybe call it `OverwriteFileIfAlreadyExistOnServer` or something like that to make it more explicit what this option will do if set to True.

maximemulder replied:
Hhhhm, the setting seemed to work on my machine, I'd be curious to know how you passed the argument (if you tested using my code or yours for instance). I think `OverwriteFileIfAlreadyExistOnServer` is a bit long and should rather belong in the documentation, but no strong opinion, either way is fine with me.

cmadjar replied:
yes, I agree, it is definitely too long. Maybe `Upload-Overwrite` would make more sense than `LORIS-Overwrite` in this case though. Just to make it clearer?

cmadjar replied:
Below is the code I am using (after having uploading the scan once):
```
dicom_upload = {
    'CandID': '300001',
    'PSCID': 'MTL001',
    'VisitLabel': 'V2',
    'IsPhantom': False,
    'LORIS-Overwrite': 'overwrite'
}
file_path = '/Users/cmadjar/Data/LORIS_imaging_data/incoming/test_api_upload/MTL001_300001_V2.tgz'
url = os.path.join(api_cred['host'], 'api', api_cred['version'], 'candidates', '300001', 'V2', 'dicoms')
try:
    response = requests.post(
        url=url,
        headers={'Authorization': 'Bearer %s' % jwtoken},
        data={'Json': json.dumps(dicom_upload)},
        allow_redirects=False,
        files={'File': (os.path.basename(file_path), open(file_path, 'rb'), 'application/tar')}

    )
    print(response)
    print(response.json())
```
When overwriting the file, I get the following response:
```
<Response [409]>
{'error': 'This file has already been uploaded.'}
```

So I m not able to overwrite the file. Not sure why it works for you but not me.

maximemulder replied:
Oh, it is because you are passing `LORIS-Overwrite: overwrite` as a JSON attribute, whereas it should be passed as an HTTP header. Try changing these lines:
```diff
-        headers={'Authorization': 'Bearer %s' % jwtoken},
+        headers={
+            'Authorization': 'Bearer %s' % jwtoken,
+            'LORIS-Overwrite': 'overwrite',
+        }
```
The parameter passed in the HTTP headers was in the original PR, I don't know what is the rationale here, so no opinion on HTTP header vs JSON attribute on my part.

cmadjar replied:
ah ben oui. Haha. OK. Not sure if the doc should be made clearer somehow? I guess it says 'optional header' but I did not understand that it was referring to the HTTP headers. Maybe it was just not clear for me and other developers would get it.

cmadjar replied:
BTW, when you call it correctly, it works ;) Glad I put the code snippet in my comment!

cmadjar commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
```diff
@@ -823,90 +823,192 @@ The JSON object is of the form:
+
+An optionnal header 'LORIS-Overwrite' with the value 'overwrite' can be used to
+overwrite an existing file.
+A successful request will be answered by a `303 See Other` response with its
```
When I tested, I received the following response, which does not corresponds to the specs in the doc here:
```
<Response [200]>
{'MriUploads': [{'MriUploadID': 125, 'Processes': []}]}
```

maximemulder replied:
This is working for me, and probably for you as well ! 303 is the redirect status code so your client is probably just following the redirect when it receives the response. If you are using `requests` to test the PR, try passing `allow_redirects=False` to the `requests.post`.

cmadjar replied:
with the `allow_redirects=False`, I did receive the 303 response.

cmadjar commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
````diff
@@ -823,90 +823,192 @@ The JSON object is of the form:
+  "ProcessType": "mri_upload",
+  "MriUploadID": 123
+}
+```
````
Here I get:
```
[Tue Oct 01 15:54:01.123601 2024] [php:notice] [pid 12396] [client ::1:51779] Cannot run command source '/Users/cmadjar/Development/Github/Loris-MRI/environment_loris' ; '/Users/cmadjar/Development/Github/Loris-MRI/uploadNeuroDB/imaging_upload_file.pl' -upload_id '127' -profile 'prod' '/Users/cmadjar/Data/LORIS_imaging_data/incoming/MTL001_300001_V2.tgz'
```

The paths are correct (even if they look unusual). Is there a config I need to set or something I may be missing?

cmadjar commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
````diff
@@ -823,90 +823,192 @@ The JSON object is of the form:
 scans, respectively.
 
+```
+POST /candidates/$CandID/$VisitLabel/dicoms
````
Is there a way to catch the apache error when file size is too big and return an appropriate error? Currently, when the file size exceeds the apache limits, we get the following message:
```
<Response [400]>
{'error': 'There should be one file uploaded per POST request.'}
```
Which is not very informative. At minimum, it should tell the user to check the apache log instead of saying there should be one file uploaded per POST

maximemulder replied:
It seems that it is very hard to know whether a file that exceeds the upload size limit was uploaded in PHP. I updated the error message in [this commit](https://github.com/aces/Loris/pull/9154/commits/39d655a9a3e012b0c6160621c8932ac1c7465f37).

cmadjar commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
````diff
@@ -823,90 +823,192 @@ The JSON object is of the form:
+    }
+  ]
+}
+```
````
Looks like the response for the get after a failed run is incorrect:
```
<Response [200]>
{'MriUploads': [{'MriUploadID': 124, 'Processes': [{'ID': 52, 'PID': 59709, 'STATE': 'SUCCESS', 'PROGRESS': 'MRI upload task (pid = 59709) failed: error code is 2', 'EXIT_CODE': 0, 'END_TIME': '2024-10-01 17:39:52'}]}]}
```

In that case, STATE should be failed and error code 2, no?

maximemulder replied:
Hhhhm, probably, this seems to be an error with the server process monitor not registering the correct exit code in the database (if that code is indeed 2). Not sure if it is related to this PR but I'll look into it.

cmadjar replied:
weird. OK. Sounds like a bug on the server process manager side then. Not sure why it is not getting the correct exit code... Maybe you could create an issue for this? Obviously, this will not get tackled in this PR.

driusan replied:
Is this thread resolved?

maximemulder replied:
Issue created (#9459). We agreed with Cécile that this is not an issue for this PR.

cmadjar reviewed (commented):
sorry for the multiple reviews in the same day but I finally got it your code working on my MCIN VM. I am able to start processes and successfully run the pipeline there 🥳 . 

While testing the process endpoints I noticed a few other things for improvements.

maximemulder commented:
> @maximemulder I re-reviewed quickly your PR. A few additional comments:
> 
>     * In the description of the PR, could you add a "Notes for existing project" with the information of the SQL patches to run?

Done !

> @maximemulder Maybe I am not seeing the right things but it does not look like you replaced all numbers by strings in the documentation, no? Including the example above.

That was the case at the beginning of the PR but I undid this change. Now the only remaining string to int in the documentation is one where the current documentation is incorrect.

> I may be wrong but I would tend to rely only on the API version provided in the API URL instead of hardcoded random API versions in the code but I will defer this one to @driusan for final decision.

This was in the original PR but I agree it should be changed, [commit here](https://github.com/aces/Loris/pull/9154/commits/085a23c0c5188c95045157723923f87337801668).

cmadjar reviewed (commented):
@maximemulder I am still having issues with overwriting an upload unfortunately. I left the code I used with the response in the comment below.

cmadjar commented:
@maximemulder looks like one of the test is failing. Other than that, I don't see any blocking comment from me and the tests were successful 👍

maximemulder commented:
Do not merge. After discussion with Cecile I will simplify the overwrite process before merging.

maximemulder commented:
I (finally) went back to apply this small simplification. Ready to merge.

driusan commented on modules/api/docs/LorisRESTAPI.md:
```diff
@@ -787,82 +787,184 @@ The JSON object is of the form:
+The JSON can optionally have a boolean `Overwrite` attribute that can be used
+to overwrite an existing file (`false` if not present).
+A successful request will be answered by a `303 See Other` response with its
+`Location` header pointing to the processes list of the new upload.
```
Shouldn't it be pointing to the new upload?

If an API user wanted the process list instead of the file itself after uploading a file, they could just append "/processes".

driusan commented on modules/api/docs/LorisRESTAPI.md:
```diff
@@ -787,82 +787,184 @@ The JSON object is of the form:
+scanning session, and as retrieved from `/candidates/$CandID/$VisitLabel/dicoms`.
 
-Only `GET` is currently supported.
+To get a list of the processes and their status for a given DICOM study previously uploaded use the following:
```
This is a different endpoint, seems to be missing a subheader.

maximemulder replied:
https://github.com/aces/Loris/pull/9154/commits/84b63983e9e4f3eec80c3940dac0e0029559d755

maximemulder replied:
https://github.com/aces/Loris/pull/9154/commits/8230e890555a37af5a6301f4240e43172988c1a6

This commit removes the documentation from the API v0.0.3. In practice, the new endpoints were already v0.0.4-only, but there are a few v0.0.4 POST routes that are added to v0.0.3 endpoints, not sure what do do here (ignore or implement method specific supported versions ?).

maximemulder replied:
Subheaders v2:

https://github.com/aces/Loris/pull/9154/commits/0c521249d1c16dd5beb965f191d4fc17c4818715

driusan commented on modules/api/docs/LorisRESTAPI.md:
````diff
@@ -787,82 +787,184 @@ The JSON object is of the form:
+GET /candidates/$CandID/$VisitLabel/dicoms/$TarName/processes
+```
+
+The response contains all `mri_upload` attempts with the specified `$tarname`. And for
````
```suggestion
The response contains all `mri_upload` attempts with the specified `$TarName`. And for
```

driusan commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
```diff
@@ -823,90 +823,192 @@ The JSON object is of the form:
 
-Like the imaging data, the DICOM data mostly lives in the `/candidates/$CandID/$Visit` 
-portion of the REST API namespace, but is defined in a separate section of this 
+Like the imaging data, the DICOM data mostly lives in the `/candidates/$CandID/$VisitLabel`
```
I just realized my above comments were made on the 0.0.3 documentation which shouldn't be changed because it's already released. Pretend I made them on the v0.0.4-dev..

maximemulder replied:
Isn't it ok to change the 0.0.3 documentation if it only clarifies the current behavior without changing it ? I would think it is ok personally.

driusan replied:
Yes, clarifying the documentation is fine, changing the behaviour (incl adding new things) should only go into the dev version. An API client that uses version $x of the API should be able to expect the same behaviour if they point to a different LORIS instance.

maximemulder replied:
So I am not sure of what you are asking me to do. Do you want the API DICOM upload endpoints to only be added to 0.0.4-dev ?

driusan replied:
Yes. (And pretend that the comments I made on them in 0.0.3 before realizing I was looking at the wrong one were made on 0.0.4-dev)

driusan replied:
Do you want me to duplicate the comments?

maximemulder replied:
No need to duplicate the comments, I am going through each of them one by one anyway.

maximemulder replied:
https://github.com/aces/Loris/commit/8230e890555a37af5a6301f4240e43172988c1a6

This commit removes the documentation from the API v0.0.3. In practice, the new endpoints were already v0.0.4-only, but there are a few v0.0.4 POST routes that are added to v0.0.3 endpoints, not sure what do do here (ignore or implement method specific supported versions ?).

driusan commented on modules/api/docs/LorisRESTAPI_v0.0.4-dev.md:
```diff
@@ -823,90 +823,192 @@ The JSON object is of the form:
-all the raw DICOM data which have been acquired for that visit. It will return an 
-object of the form:
+A GET request to `/candidates/$CandID/$VisitLabel/dicoms` will return a JSON object of
+all the DICOM archives (A.K.A. one DICOM archive per `StudyInstanceUID`) which have been
```
It's not a huge deal but I don't think the "a.k.a" parenthesis belongs in technical documentation. It sounds very informal.

driusan commented on modules/api/php/endpoints/candidate/visit/dicom/processes.class.inc:
```diff
@@ -0,0 +1,255 @@
+                $mriupload->getUploadLocation()
+            );
+        } catch (\LorisException $e) {
+            error_log($e->getMessage());
```
This should be using the PSR-3 logger for exceptions so that the config setting is honoured.

maximemulder replied:
https://github.com/aces/Loris/pull/9154/commits/6f3651d8fa430ec4d3ad51242995db097922e62e

driusan commented on modules/api/php/endpoints/candidate/visit/dicom/processes.class.inc:
```diff
@@ -0,0 +1,255 @@
+        );
+
+        $lorispath = $request->getQueryParams()['lorispath'] ?? '';
+        $newpath   = $lorispath . '/' .$processid;
```
Concatenating strings with user input is very risky. I don't think there's a direct attack possible the string is just going into a response, but it's possible to construct arbitrary urls that escape the module if an attacker knows what you're going to do with it.

1. Use Utility::pathJoin if you need to join path.
2. Don't assume lorispath is reliable, it comes from a get parameter in an htaccess rewrite. Use the BaseURL instead. Also you removed lorispath from the request object in a55cb25bf1daa167f0b1055a0ae272f67667a741

maximemulder replied:
Whoops, this was old code I forgot to update and it seems there was indeed a bug here. Fixed in this commit:
https://github.com/aces/Loris/pull/9154/commits/bfb2873811046cdd4e1c5b5b1550b4800fbce82d
(it still uses concatenation but that should be enough IMO)

driusan commented on modules/api/php/endpoints/candidate/visit/dicoms.class.inc:
```diff
@@ -153,6 +171,170 @@ class Dicoms extends Endpoint implements \LORIS\Middleware\ETagCalculator
+            return new \LORIS\Http\Response\JSON\BadRequest(
+                'No file found, there should be one file uploaded per request. '
+                    . 'If a file was uploaded, check that it is not larger than '
+                    . 'the max file size PHP setting or take a look at the '
```
It is an API user who sees this message, not a LORIS sysadmin. They can not check the max file size PHP setting or look a the Apache logs.

maximemulder replied:
Changed the message.
https://github.com/aces/Loris/pull/9154/commits/59437dec438d0ea7a02d7ebfef988b737ad24885

It is less specific but should still be enough for a LORIS dev to fix the error, which is important based on how our tests with Cécile went.

driusan commented on modules/api/php/endpoints/candidate/visit/dicoms.class.inc:
```diff
@@ -153,6 +171,170 @@ class Dicoms extends Endpoint implements \LORIS\Middleware\ETagCalculator
+                $is_phantom,
+            );
+        } catch (\Throwable $e) {
+            error_log($e->getMessage());
```
Should use the logger

maximemulder replied:
https://github.com/aces/Loris/pull/9154/commits/0488bb032d190bc9fdb229cf6cd6d3057e418c5a

driusan commented on modules/api/php/views/visit/dicom/processes.class.inc:
```diff
@@ -0,0 +1,72 @@
+    {
+        return [
+            'MriUploads' => array_map(
+                self::_formatMRIUploads(...),
```
I am surprised this isn't a syntax error with nothing after the `...`, can you explain what you're trying to do here?

maximemulder replied:
I am using the [first class callable syntax](https://www.php.net/manual/en/functions.first_class_callable_syntax.php). If I remember correctly the function used to be passed as a string, but this modern syntax allows to not use a string (which is nice IMO).

driusan replied:
That makes sense. I was unaware of the syntax and thought it was a splat operator missing a value.

driusan commented on modules/server_processes_manager/php/mriuploadserverprocess.class.inc:
```diff
@@ -52,25 +52,25 @@ class MriUploadServerProcess extends AbstractServerProcess
 
     /**
-     * Root of the MRI files directory.
+     * The value of MRICodePath config setting.
```
The changes to this file don't seem to be related to the API and should be moved to a different PR so someone more familiar with the server_processes_manager can review them.

driusan commented on src/Data/Models/MRIUploadDTO.php:
```diff
@@ -0,0 +1,194 @@
+ */
+class MRIUploadDTO implements \LORIS\Data\DataInstance
+{
+    private $mri_upload_id;
```
Not necessary, but these could just use constructor property promotion.

maximemulder replied:
Is this ok ?
https://github.com/aces/Loris/pull/9154/commits/4d6eeb1e2c2d4d99a8590538de4c88521783695b

driusan commented on src/Http/Response/JSON/Accepted.php:
```diff
@@ -0,0 +1,40 @@
+<?php declare(strict_types=1);
```
Surprised this doesn't exist but if you move it to a different PR it could be easily merged.

driusan commented on src/Http/Response/JSON/SeeOther.php:
```diff
@@ -0,0 +1,45 @@
+<?php declare(strict_types=1);
```
This could also be easily merged if it were separated out.

driusan reviewed (changes requested):
various comments

maximemulder commented:
@cmadjar One of your comments was that you would like to the DICOM API upload to re-use an existing MRI upload if there is one. Is the following commit good for you ?

https://github.com/aces/Loris/pull/9154/commits/d3eb5c5ce1ef970c8a33706fb8b0e3d5b62fb611

I notably do not update the upload timestamp as the DICOM archive has its own timestamp and the MRI upload may be referenced by other, older, data. But I can change that if you want to. I also only re-use an existing MRI upload if its parameters are similar (upload location, uploader id...).

cmadjar reviewed (approved):
LGTM. Thanks for the last change!

maximemulder commented:
@driusan I believe I answered all the issues that were raised by Cécile (who approved the PR) and you. I also updated the 0.0.4-dev Swagger schema to match the markdown documentation. You can review the last 7 commits commit-by-commit to see the changes.

cmadjar commented:
@driusan this is in your court now. I approved the PR a while back and tested. I believe @maximemulder addressed everything so we just need a final look from you now. Assigned the PR to you.

driusan commented on php/libraries/MRIUpload.class.inc:
```diff
@@ -0,0 +1,80 @@
+     */
+    public function getProcesses(): array
+    {
+        $rows = \NDB_Factory::singleton()->database()->pselectCol(
```
Database connections should come from the LorisInstance->getDatabaseConnection. You can either pass it (or the Database connection itself) as a parameter or inject it as a constructor parameter. (I have no preference)

driusan reviewed (changes requested):
Please revert the changes to the v0.0.3 documentation (if it's inconsistent with the code you can open an issue) and remove the references to NDB_Factory->database() (either inject the database class or get it from a LorisInstance object), then I'll merge.

maximemulder commented:
@driusan See the three latest commits for your requested changes:
- Rebased on main and fixed new type error.
- Add the LORIS instance to some classes to prevent using the database singleton.
- Revert the documentation fix in the 0.0.3 and 0.0.4-dev documentation (I still let the formatting changes in 0.0.3 to stay consistent with the 0.0.4-dev documentation).

maximemulder commented:
Re-rebased on main and reset 0.0.3 documentation to main.
